### PR TITLE
Ajo redis/LUA

### LINKED ISSUE #1: Change pg to redis
# PG to redis

### Data structure
Model stored in keys, leaderboard in sorted set.
We use keys instead of a hash to benefit from redis' ttl system.

```bash
# [
#   {id: 111, name: "Zymna", count: 7}
#   {id: 222, name: "Axl", count: 4}
# ]
> incrby 111:count 7
> incrby 222:count 4

# leaderboard implemented as a sorted set
> zincrby leaderboard 7 "Zymna"
> zincrby leaderboard 4 "Axl"

# expiring award operations are stored in :{type}
> set 111:daily 1 ex 86400 # daily ttl
> set 111:weekly 1 ex 604800 # weekly ttl
```

### Pseudo code
EDIT: keeping the pseudo code for reference, but will be implemented as LUA.

For performance, methods will call redis as few as possible, making use of `pipe`.
All commands changing redis data will acquire locks for the impacted users.

```py
# assume an existing redis connection
def verajo(user):
  return redis.get("{user.id}:count")

def ajo(user):
  return redis.incrby("{user.id}:count 1")

# top 12 scores
def leaderboard():
  return redis.zrange("leaderboard 0 11 rev withscores")

# all award expiring operations (daily/weekly) are the same
def award(user, expire, type, award):
  locked, ttl = redis.pipe(
    _set("{user.id}:lock 1 NX EX 1")
    _ttl("{user.id}:{type}")
  )
  if not locked:
    err("locked")

  # user needs to wait until ttl
  if ttl > 0:
    return ttl

  redis.pipe(
    _set("{user.id}:{type} 1 EX {expire}")
    _incrby("{user.id}:count {award}")
    _zincrby("leaderboard {award} {user.name}")
    _del("{user.id}:lock")
  )

def daily(user):
  return self.award(user, "daily", 24 * 3600, DAILY)

def weekly(user):
  return self.award(user, "weekly", 24 * 3600 * 7, WEEKLY)

def gamble(user, amount):
  if amount < 1:
    err("too few")

  locked, current = redis.pipe(
    _set("{user.id}:lock 1 NX EX 1")
    _get("{user.id}:count")
  )
  if not locked:
    err("locked")

  if amount > current:
    err("too much")

  # 25% to win something
  if randrange(0, 3) == 1:
    # from 1 to 250%
    pct_win = randrange(1, 100) / 40
    change = ceil(pct_win * amount)
  else:
    change = -amount

  # update local count, lb and clear lock
  redis.pipe(
    _incrby("{user.id}:count {change}")
    _zincrby("leaderboard {change} {user.name}")
    _del("{user.id}:lock")
  )

  return change

def pay(from_user, to_user, amount):
  if amount < 1:
    err("too few")

  from_locked, to_locked, from_current = redis.pipe(
    _set("{from_user.id}:lock 1 NX EX 1")
    _set("{to_user.id}:lock 1 NX EX 1")
    _get("{from_user.id}:count")
  )
  if not from_locked or not to_locked:
    err("locked")

  if amount > from_current:
    err("too much")

  rem_amount = -amount
  redis.pipe(
    _incrby("{from_user.id}:count {rem_amount}")
    _zincrby("leaderboard {rem_amount} {from_user.name}")
    _incrby("{to_user.id}:count {amount}")
    _zincrby("leaderboard {amount} {to_user.name}")
    _del("{from_user.id}:lock", "{to_user.id}:lock")
  )

def discombobulate(from_user, to_user, amount):
  if amount < 1:
    err("too few")

  from_locked, to_locked, from_current, to_current = redis.pipe(
    _set("{from_user.id}:lock 1 NX EX 1")
    _set("{to_user.id}:lock 1 NX EX 1")
    _get("{from_user.id}:count", "{to_user.id}:count")
  )
  if not from_locked or not to_locked:
    err("locked")

  if amount > from_current:
    err("too much")

  # need to offer 35% of what the other has at least
  if amount <= ((35/100) * to_current):
    err("not enough")

  # discombobulate dmg is 69 to 200% of offered
  pct_dmg = randrange(69, 200)
  dmg = round(amount * (pct_dmg / 100))

  rem_amount = -amount
  redis.pipe(
    _incrby("{from_user.id}:count {rem_amount}")
    _zincrby("leaderboard {rem_amount} {from_user.name}")
    _incrby("{to_user.id}:count {dmg}")
    _zincrby("leaderboard {dmg} {to_user.name}")
    _del("{from_user.id}:lock", "{to_user.id}:lock")
  )

  return dmg
```

### LINKED ISSUE #10: Daily/weekly did not work
Maybe a lag/cache/concurrency issue?

ajobug

### PR DESCRIPTION
**Pull request by Zymna**
Missing todo:
- [x] include redis 6.2+
- [x] remove pg from docker
- [x] remove pg/alembic files
- [x] remove unnecessary models
- [x] script load at bootstrap (`redis-cli -x script load < {script.lua}`)
- [x] use evalsha + redis module in python

Closes #1
Closes #10

### CODE REVIEW

**Comment by axl89** [comment]
Should we remove `ormar` dependency as well? It's used somewhere:

In manager.py

```python
from ormar import NoMatch
...
            except NoMatch:
                self._cache[user.id] = await GarlicUser(user=user.id, name=f"{user.name}#{user.discriminator}").save()

```

In user.py:

```python
from ormar import BigInteger, DateTime, Model, String
...
    user: int = BigInteger(primary_key=True, autoincrement=False)
    name: str = String(max_length=255, nullable=False)
    count: int = BigInteger(default=0)

```

**Comment by Zymna** [comment]
`GarlicUser` should not be necessary anymore, I believe it represents the db model which we don't have anymore.

File: pyproject.toml [hunk 1]
```diff
@@ -10,12 +10,10 @@ repository = "https://github.com/lastlife/ajobot"
 python = "^3.10"
 taskipy = "^1.9.0"
 disnake = "^2.4.0"
-asyncpg = "^0.25.0"
 loguru = "^0.5.3"
-alembic = "^1.7.5"
-ormar = "^0.10.23"
 python-dotenv = "^0.19.2"
 psycopg2-binary = "^2.9.3"
```
Comment (Zymna) [inline pyproject.toml:15]: @axl89 should this be removed? Looks like pg

**Comment by Zymna** [comment]
Changed the system to remove the count keys which are useless: we can use the leaderboard directly.
In short:
```diff
- GET 111:count
+ ZSCORE leaderboard Lv
```

**Pull request merged**